src/core/service.c

```diff
@@ -1358,6 +1382,81 @@ static int service_collect_fds(
+                if (S_ISSOCK(st.st_mode)) {
+                        int socketfd;
+                        struct sockaddr_un addr;
+                        socklen_t sa_len;
+                        socketfd = socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC, 0);
+                        if (socketfd < 0) {
+                                log_unit_error_errno(UNIT(s), errno, "Failed to create socket for %s: %m", f->path);
+                                continue;
+                        }
+                        r = sockaddr_un_set_path(&addr, f->path);
+                        if (r < 0) {
+                                log_unit_error_errno(UNIT(s), errno, "Failed to fill sockaddr for %s: %m", f->path);
+                                continue;
+                        }
+                        sa_len = r;
+                        f->fd = connect(socketfd, (struct sockaddr*)&addr, sa_len) >= 0 ? socketfd : -1;
+                        if (f->fd < 0) {
+                                log_unit_error_errno(UNIT(s), errno, "Failed to connect to socket sun_path %s sun_family %d: %m", addr.sun_path, addr.sun_family);
+                                continue;
+                        }
+                        log_unit_debug(UNIT(s), "[service_collect_fds] socket %s connected (fd=%d)", f->path, f->fd);
+                        ++n_openfile_fds;
+                } else {
+                        f->fd = open(f->path, f->flags);
+                        if (f->fd < 0) {
+                                log_unit_error_errno(UNIT(s), errno, "Failed to open file %s: %m", f->path);
+                                continue;
+                        }
+                        log_unit_debug(UNIT(s), "[service_collect_fds] file %s opened (fd=%d)", f->path, f->fd);
+                        ++n_openfile_fds;
+                }
+
+        }
+        if (n_openfile_fds > 0) {
+                size_t n_fds;
+                char **nl;
+                int *t;
+
+                t = reallocarray(rfds, rn_socket_fds + s->n_fd_store + n_openfile_fds, sizeof(int));
+                if (!t)
+                        return -ENOMEM;
+
+                rfds = t;
+
+                nl = reallocarray(rfd_names, rn_socket_fds + s->n_fd_store + n_openfile_fds + 1, sizeof(char *));
+                if (!nl)
+                        return -ENOMEM;
+
+                rfd_names = nl;
+                n_fds = rn_socket_fds;
+
+                LIST_FOREACH(open_files, f, s->open_files) {
+                        if (f->fd >= 0) {
+                                rfds[n_fds] = f->fd;
+                                rfd_names[n_fds] = strdup(strempty(f->fdname));
+                                if (!rfd_names[n_fds])
+                                        return -ENOMEM;
+
+                                rn_storage_fds++;
+                                n_fds++;
+                        }
+                }
+
+                rfd_names[n_fds] = NULL;
+        }
```

Comment:
Why do this here after parsing all the `OpenFile=` settings? I would remove the `fd` field from `OpenFile=` and just dynamically grow `rfds` as you're looping over the `OpenFile=` settings. No need to do it in a separate loop. You can use `GREEDY_REALLOC()` to grow the array dynamically in an efficient way.